File: src/main/scala/info/crlog/higgs/HiggsClient.scala
package info.crlog.higgs

//use _root_. to make package resolution absolute, otherwise scala prepends info.crlog.higgs

import _root_.java.util.concurrent.Executors
import _root_.java.net.InetSocketAddress
import org.jboss.netty.channel.socket.nio.NioClientSocketChannelFactory
import org.jboss.netty.bootstrap.ClientBootstrap
import protocol._
import boson.Publisher

/**
 * A simple client interface to encapsulate the Netty NIO connection for a client request
 * @author Courtney Robinson <courtney@crlog.info> @ 31/01/12
 * @param host the host name to connect to
 * @param port the port on the host
 * @param decoder the class type of the protocol's decoder which must inherit from HiggsDecoder, see http://stackoverflow.com/a/4870084/400048 for   concise explanation of Scala's upper bound syntax
 * @param encoder the class type of the protocol's encoder
 * @param clientHandler the client handler which implements the "business" logic for the protocol
 */

class HiggsClient(host: String, port: Int,
                  decoder: Class[_ <: HiggsDecoder],
                  encoder: Class[_ <: HiggsEncoder],
                  clientHandler: Class[_ <: HiggsPublisher]
                   ) {

  val channelFactory = new NioClientSocketChannelFactory(Executors.newCachedThreadPool(), Executors.newCachedThreadPool())
  // Configure the client.
  val bootstrap = new ClientBootstrap(channelFactory);

  // Set up the event pipeline factory.
  bootstrap.setPipelineFactory(new ClientPipelineFactory(decoder, encoder, clientHandler));

  // Make a new connection.
  val connectFuture = bootstrap.connect(new InetSocketAddress(host, port));

  // Wait until the connection is made successfully.
  val channel = connectFuture.awaitUninterruptibly().getChannel();

  //Get the handler instance
  val handler = channel.getPipeline.getLast.asInstanceOf[HiggsPublisher]
  //  /**
  //   * Adds a message listener that will be notified when messages are received
  //   */
  //  def addListener(listener: MessageListener) = {
  //    handler.addListener(listener)
  //  }

  def shutdown() = {
    // Shut down all thread pools to exit.
    bootstrap.releaseExternalResources();
  }
}

File: src/main/java/info/crlog/higgs/HiggsClient.java
package info.crlog.higgs;

import info.crlog.higgs.messaging.HiggsDecoder;
import info.crlog.higgs.messaging.HiggsEncoder;
import io.netty.bootstrap.ClientBootstrap;
import io.netty.channel.Channel;
import io.netty.channel.ChannelFuture;
import io.netty.channel.socket.nio.NioClientSocketChannelFactory;
import io.netty.example.telnet.TelnetClient;
import java.net.InetSocketAddress;
import java.util.concurrent.Executors;

/**
 * Simple SSL chat client modified from {@link TelnetClient}.
 */
public abstract class HiggsClient {

    protected int port = 2012;
    protected String host = "localhost";
    protected ClientBootstrap bootstrap;
    protected ClientPipelineFactory pipeline;
    protected ClientHandler handler;
    protected HiggsDecoder decoder;
    protected HiggsEncoder encoder;
    protected boolean useSSL;
    protected Channel channel;
    protected ChannelFuture future;

    public HiggsClient(String host, int port) {
        this.host = host;
        this.port = port;
        handler = new ClientHandler(useSSL);
    }

    /**
     * Release external resources
     */
    public void shutdown() {
        bootstrap.releaseExternalResources();
    }

    /**
     * Connect to the host:port provided
     *
     * @return true IF AND ONLY IF the client is not already connected and the
     * connection attempt was successful. In all other cases false is returned
     * and the connection can immediately be retried
     */
    protected boolean connect() {
        if (decoder == null || encoder == null) {
            throw new EncoderDecoderInitializationException();
        }
        if (bootstrap == null || channel == null || !channel.isConnected()) // Configure the client.
        {
            bootstrap = new ClientBootstrap(
                    new NioClientSocketChannelFactory(
                    Executors.newCachedThreadPool()));
            // Configure the pipeline factory.
            bootstrap.setPipelineFactory(new ClientPipelineFactory(handler,
                    encoder,
                    decoder,
                    useSSL));
            // Start the connection attempt.
            future = bootstrap.connect(new InetSocketAddress(
                    host,
                    port));
            // Wait until the connection attempt succeeds or fails.
            channel = future.awaitUninterruptibly().getChannel();
            if (!future.isSuccess()) {
                bootstrap.releaseExternalResources();
                channel = null;
                bootstrap = null;
                return false;
            } else {
                return true;
            }
        } else {
            return false;
        }
    }

    /**
     * Perform internal initialization. Automatically invoked by the parent
     * HiggsServer on construction. Use to initialize the encoder,decoder for
     * e.g.
     */
    protected abstract void initialize();
}
